File: agents/adapters/llm/src/llm_adapter/router.py
# SPDX-License-Identifier: Apache-2.0
"""CapabilityRouter — maps capability names to handlers and JSON Schema bytes."""

from __future__ import annotations

import json

from llm_adapter.config import ProviderConfig
from llm_adapter.handler import ChatCompletionHandler

_CHAT_INPUT_SCHEMA: dict[str, object] = {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "required": ["prompt"],
    "properties": {
        "prompt": {"type": "string", "minLength": 1},
        "system": {"type": "string"},
        "temperature": {"type": "number", "minimum": 0, "maximum": 2},
        "max_tokens": {"type": "integer", "minimum": 1},
    },
    "additionalProperties": False,
}

_CHAT_OUTPUT_SCHEMA: dict[str, object] = {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "type": "object",
    "required": ["response"],
    "properties": {"response": {"type": "string"}},
}

_INPUT_BYTES: bytes = json.dumps(_CHAT_INPUT_SCHEMA).encode()
_OUTPUT_BYTES: bytes = json.dumps(_CHAT_OUTPUT_SCHEMA).encode()


class CapabilityRouter:
    """Routes capability names to handlers and exposes their JSON Schemas.

    Built once at adapter startup from ``ProviderConfig``. Immutable after
    construction — all state is read-only after ``__init__``.
    """

    def __init__(self, config: ProviderConfig) -> None:
        """Initialise the router with a single chat_completion capability.

        Args:
            config: Validated provider configuration used to build the handler.
        """
        self._handlers: dict[str, ChatCompletionHandler] = {
            "chat_completion": ChatCompletionHandler(config),
        }
        self._schemas: dict[str, tuple[bytes, bytes]] = {
            "chat_completion": (_INPUT_BYTES, _OUTPUT_BYTES),
        }

    def dispatch(self, capability_name: str) -> ChatCompletionHandler:
        """Return the handler for a named capability.

        Args:
            capability_name: Capability name from ``ExecuteCapabilityRequest``.

        Returns:
            The ``ChatCompletionHandler`` registered for this capability.

        Raises:
            KeyError: When ``capability_name`` is not registered.
        """
        return self._handlers[capability_name]

    def get_schema(self, capability_name: str) -> tuple[bytes, bytes]:
        """Return the (input_schema, output_schema) JSON bytes for a capability.

        Args:
            capability_name: Registered capability name.

        Returns:
            Tuple of (input_schema_json_bytes, output_schema_json_bytes).

        Raises:
            KeyError: When ``capability_name`` is not registered.
        """
        return self._schemas[capability_name]

    def capability_names(self) -> list[str]:
        """Return all registered capability names."""
        return list(self._handlers.keys())

File: agents/adapters/llm/internal/domain/router.go
// SPDX-License-Identifier: Apache-2.0

// Package domain implements the capability routing and request handling for the
// llm-adapter: a CapabilityRouter maps capability names to a ChatCompletionHandler
// and the declared JSON Schemas, and the handler validates input, streams provider
// tokens as PROGRESS events, and emits exactly one terminal event (canvas M7.P.4).
package domain

import (
	"github.com/zynax-io/zynax/agents/adapters/llm/internal/config"
	"github.com/zynax-io/zynax/agents/adapters/llm/internal/provider"
)

// capabilitySchema holds the declared JSON Schemas and description for a single
// capability, returned verbatim by GetCapabilitySchema.
type capabilitySchema struct {
	description  string
	inputSchema  string
	outputSchema string
}

// CapabilityRouter maps capability names to a handler and the declared JSON
// Schemas. It is built once from AdapterConfig at startup and is immutable after
// construction — all fields are read-only after New (canvas: stateless adapter).
type CapabilityRouter struct {
	handlers map[string]*ChatCompletionHandler
	schemas  map[string]capabilitySchema
}

// NewRouter builds a CapabilityRouter from a validated AdapterConfig and a single
// shared Provider. Each declared capability is bound to a ChatCompletionHandler
// that validates against the capability's input schema before invoking p.
func NewRouter(cfg *config.AdapterConfig, p provider.Provider) (*CapabilityRouter, error) {
	handlers := make(map[string]*ChatCompletionHandler, len(cfg.Capabilities))
	schemas := make(map[string]capabilitySchema, len(cfg.Capabilities))
	for _, c := range cfg.Capabilities {
		h, err := newChatCompletionHandler(p, c.InputSchemaJSON)
		if err != nil {
			return nil, err
		}
		handlers[c.Name] = h
		schemas[c.Name] = capabilitySchema{
			description:  c.Description,
			inputSchema:  c.InputSchemaJSON,
			outputSchema: c.OutputSchemaJSON,
		}
	}
	return &CapabilityRouter{handlers: handlers, schemas: schemas}, nil
}

// Dispatch returns the handler registered for capability name and whether it
// exists. The boolean lets the caller map an unknown capability to NOT_FOUND.
func (r *CapabilityRouter) Dispatch(name string) (*ChatCompletionHandler, bool) {
	h, ok := r.handlers[name]
	return h, ok
}

// Schema returns the declared description and JSON Schemas for capability name
// and whether it exists.
func (r *CapabilityRouter) Schema(name string) (description, inputSchema, outputSchema string, ok bool) {
	s, found := r.schemas[name]
	if !found {
		return "", "", "", false
	}
	return s.description, s.inputSchema, s.outputSchema, true
}
